infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureDetector.java

```diff
@@ -0,0 +1,332 @@
+            }
+        }
+
+        // We received responses from all the members.
+        if (responsiveNodes.equals(membersSet)) {
+            // We can try to scale back the network latency time only if all pings received.
+            period = Math.max(initPeriodDuration, (period - periodDelta));
+            // If there are no ping failures and no out of phase epochs then end round.
+            if (Arrays.stream(expectedEpoch).filter(l -> l != INVALID_EPOCH).count() == 0) {
+                return false;
+            }
+
+        } else {
+            if (newPeriod != period && newPeriod != 0) {
+                period = newPeriod;
+                tuneRoutersResponseTimeout(membersSet, period);
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Each PollRound executes "failureThreshold" number of iterations and generates the
+     * poll report.
+     *
+     * @return Poll Report with detected failed nodes and out of phase epoch nodes.
+     */
+    private PollReport pollRound() {
+
+        Set<String> membersSet = new HashSet<>(Arrays.asList(members));
+        boolean failuresDetected = true;
+
+        // At the start of the round reset all response counters.
+        resetResponses();
+
+        // In each iteration we poll all the servers in the members list.
+        for (int iteration = 1; iteration <= failureThreshold; iteration++) {
+
+            // Ping all nodes and await their responses.
+            pollOnceAsync();
+
+            // Collect responses and increment response counters for successful pings.
+            collectResponsesAndVerifyEpochs(iteration);
+
+            failuresDetected = isFailurePresent(membersSet, iteration);
+            if (!failuresDetected) {
+                break;
+            }
+
+            try {
+                Thread.sleep(interIterationInterval);
+            } catch (InterruptedException e) {
+                log.warn("pollRound: Sleep interrupted.");
+            }
+
+        }
+
+        Map<String, Long> outOfPhaseEpochNodes = new HashMap<>();
+        Set<String> failed = new HashSet<>();
+
+        if (failuresDetected) {
+            // Check all responses and collect all failures.
+            for (int i = 0; i < members.length; i++) {
+                if (responses[i] != failureThreshold) {
+                    failed.add(members[i]);
+                }
+                if (expectedEpoch[i] != INVALID_EPOCH) {
+                    outOfPhaseEpochNodes.put(members[i], expectedEpoch[i]);
+                    expectedEpoch[i] = INVALID_EPOCH;
+                }
+
+                // Reset the timeout of all the failed nodes to the max value to set a longer
+                // timeout period to detect their response.
+                tuneRoutersResponseTimeout(failed, maxPeriodDuration);
+            }
+        } else {
+            // Need to tune back if no failures.
+            tuneRoutersResponseTimeout(membersSet, initPeriodDuration);
```

Comment:
Does this mean that you will increase timeout period for some time and if all nodes respond you will drop it to initPeriodPeriodDuration and start again in the new round . Did we decide on that ? Also need to log when you tune the response timeout. and log the reason why you did it.

Reply:
We had decided to scale back gradually. I have made the change.
I have added the logging.